File: src/go/maintained.go
package oreslocks

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"sync"
	"time"
)

// LeaseMaintenanceOptions controls renewal while a Fiducia lease encloses a
// PostgreSQL advisory-lock transaction.
type LeaseMaintenanceOptions struct {
	// RenewInterval must be positive and no greater than half of AcquireOptions.TTL.
	RenewInterval time.Duration
}

// DefaultLeaseMaintenanceOptions leaves a 40-second margin inside the default
// 60-second Fiducia TTL.
func DefaultLeaseMaintenanceOptions() LeaseMaintenanceOptions {
	return LeaseMaintenanceOptions{RenewInterval: 20 * time.Second}
}

// Validate rejects unsafe timing before either coordination layer is acquired.
func (o LeaseMaintenanceOptions) Validate(key LockKey, acquire AcquireOptions, wait bool) error {
	switch {
	case acquire.TTL <= 0:
		return invalidPlan(key, "fiducia lease TTL must be greater than zero")
	case o.RenewInterval <= 0:
		return invalidPlan(key, "fiducia renewal interval must be greater than zero")
	case o.RenewInterval > acquire.TTL/2:
		return invalidPlan(key, fmt.Sprintf(
			"fiducia renewal interval %d ms is unsafe for TTL %d ms; it must be no greater than half the TTL",
			o.RenewInterval.Milliseconds(), acquire.TTL.Milliseconds(),
		))
	case acquire.WaitTimeout < 0:
		return invalidPlan(key, "PostgreSQL advisory-lock wait timeout must not be negative")
	case wait && acquire.RetryInterval <= 0:
		return invalidPlan(key, "PostgreSQL advisory-lock retry interval must be greater than zero when wait is enabled")
	default:
		return nil
	}
}

// MaintainedXactGuarded is what maintained work receives. Every protected SQL
// statement must use Tx. The supplied work context is canceled with the
// renewal failure as its cause so cooperative work can stop promptly.
type MaintainedXactGuarded struct {
	Guarded
	Tx *sql.Tx
}

// WithMaintainedXactLock holds a Fiducia lease around one PostgreSQL
// transaction-scoped advisory lock.
//
// Waiting is implemented with pg_try_advisory_xact_lock polling so renewal
// loss can interrupt the wait. The lease is renewed periodically while lock
// acquisition and work are pending, and one final renewal is mandatory before
// Commit. Any failure before commit rolls the transaction back, then releases
// the outer grant while preserving cleanup-error precedence.
func WithMaintainedXactLock(
	ctx context.Context,
	key LockKey,
	wait bool,
	acquire AcquireOptions,
	maintenance LeaseMaintenanceOptions,
	lease Lease,
	db *sql.DB,
	work func(context.Context, MaintainedXactGuarded) error,
) error {
	if lease == nil {
		return invalidPlan(key, "a Fiducia lease authority is required")
	}
	if db == nil {
		return invalidPlan(key, "a PostgreSQL database is required")
	}
	if work == nil {
		return invalidPlan(key, "maintained transaction work must not be nil")
	}
	if err := maintenance.Validate(key, acquire, wait); err != nil {
		return err
	}

	grant, err := acquireLease(ctx, key, wait, acquire, lease)
	if err != nil {
		return err
	}
	inner := runMaintainedXact(ctx, key, wait, acquire, maintenance, lease, grant, db, work)
	return settle(ctx, key, lease, grant, inner)
}

// WithMaintainedBoth is the blocking both-layer path with package defaults.
func WithMaintainedBoth(
	ctx context.Context,
	key LockKey,
	lease Lease,
	db *sql.DB,
	work func(context.Context, MaintainedXactGuarded) error,
) error {
	return WithMaintainedXactLock(
		ctx,
		key,
		true,
		DefaultAcquireOptions(),
		DefaultLeaseMaintenanceOptions(),
		lease,
		db,
		work,
	)
}

func runMaintainedXact(
	ctx context.Context,
	key LockKey,
	wait bool,
	acquire AcquireOptions,
	maintenance LeaseMaintenanceOptions,
	lease Lease,
	grant LeaseGrant,
	db *sql.DB,
	work func(context.Context, MaintainedXactGuarded) error,
) error {
	workCtx, maintainer := startLeaseMaintainer(ctx, lease, grant, acquire.TTL, maintenance.RenewInterval)

	tx, err := db.BeginTx(workCtx, nil)
	if err != nil {
		maintained := maintainer.stop()
		if maintained != nil {
			return cleanupFailure(maintained, dbErr(key, StepPgBegin, err))
		}
		if cause := maintenanceCause(workCtx); cause != nil {
			return cause
		}
		return dbErr(key, StepPgBegin, err)
	}

	inner := acquireMaintainedXactLock(workCtx, key, wait, acquire, tx)
	if inner == nil {
		if err := work(workCtx, MaintainedXactGuarded{
			Guarded: Guarded{Key: key, Grant: &grant},
			Tx:      tx,
		}); err != nil {
			inner = workErr(key, err)
		}
	}

	if maintained := maintainer.stop(); maintained != nil {
		inner = cleanupFailure(maintained, inner)
	}
	if inner == nil && ctx.Err() != nil {
		inner = workErr(key, ctx.Err())
	}
	if inner == nil {
		_, inner = renewChecked(ctx, lease, grant, acquire.TTL)
	}
	if inner != nil {
		return rollbackMaintained(tx, key, inner)
	}

	if err := tx.Commit(); err != nil {
		return dbErr(key, StepPgCommit, err)
	}
	return nil
}

func acquireMaintainedXactLock(
	ctx context.Context,
	key LockKey,
	wait bool,
	acquire AcquireOptions,
	tx *sql.Tx,
) error {
	started := time.Now()
	for {
		var acquired bool
		err := tx.QueryRowContext(ctx, "SELECT pg_try_advisory_xact_lock($1)", key.Advisory()).Scan(&acquired)
		if err != nil {
			if cause := maintenanceCause(ctx); cause != nil {
				return cause
			}
			step := StepPgTryAdvisoryXactLock
			if wait {
				step = StepPgAdvisoryXactLock
			}
			return dbErr(key, step, err)
		}
		if acquired {
			return nil
		}
		if !wait {
			return contention(key, StepPgTryAdvisoryXactLock)
		}

		elapsed := time.Since(started)
		if elapsed >= acquire.WaitTimeout {
			return timeout(key, StepPgAdvisoryXactLock, acquire.WaitTimeout.Milliseconds())
		}
		remaining := acquire.WaitTimeout - elapsed
		delay := acquire.RetryInterval
		if remaining < delay {
			delay = remaining
		}
		timer := time.NewTimer(delay)
		select {
		case <-ctx.Done():
			if !timer.Stop() {
				<-timer.C
			}
			if cause := maintenanceCause(ctx); cause != nil {
				return cause
			}
			return workErr(key, ctx.Err())
		case <-timer.C:
		}
	}
}

func renewChecked(
	ctx context.Context,
	lease Lease,
	original LeaseGrant,
	ttl time.Duration,
) (LeaseGrant, error) {
	renewed, err := lease.Renew(ctx, original, ttl)
	if err != nil {
		return LeaseGrant{}, tagStep(err, StepFiduciaRenew)
	}

	var changed string
	switch {
	case renewed.Key != original.Key:
		changed = "key"
	case renewed.Holder != original.Holder:
		changed = "holder"
	case renewed.FencingToken != original.FencingToken:
		changed = "fencing token"
	}
	if changed != "" {
		return LeaseGrant{}, newError(
			KindLostLease,
			original.Key,
			StepFiduciaRenew,
			"fiducia renewal changed the grant "+changed+"; fenced authority cannot be proven",
			nil,
		)
	}
	return renewed, nil
}

func rollbackMaintained(tx *sql.Tx, key LockKey, inner error) error {
	if err := tx.Rollback(); err != nil && !errors.Is(err, sql.ErrTxDone) {
		return cleanupFailure(dbErr(key, StepPgRollback, err), inner)
	}
	return inner
}

func maintenanceCause(ctx context.Context) error {
	cause := context.Cause(ctx)
	var lockErr *Error
	if errors.As(cause, &lockErr) {
		return lockErr
	}
	return nil
}

type leaseMaintainer struct {
	stopOnce sync.Once
	stopCh   chan struct{}
	doneCh   chan struct{}
	cancel   context.CancelCauseFunc

	mu      sync.Mutex
	failure error
}

func startLeaseMaintainer(
	parent context.Context,
	lease Lease,
	grant LeaseGrant,
	ttl time.Duration,
	interval time.Duration,
) (context.Context, *leaseMaintainer) {
	workCtx, cancel := context.WithCancelCause(parent)
	maintainer := &leaseMaintainer{
		stopCh: make(chan struct{}),
		doneCh: make(chan struct{}),
		cancel: cancel,
	}
	go maintainer.run(parent, lease, grant, ttl, interval)
	return workCtx, maintainer
}

func (m *leaseMaintainer) run(
	ctx context.Context,
	lease Lease,
	grant LeaseGrant,
	ttl time.Duration,
	interval time.Duration,
) {
	defer close(m.doneCh)
	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-m.stopCh:
			return
		case <-ctx.Done():
			return
		case <-ticker.C:
			if _, err := renewChecked(ctx, lease, grant, ttl); err != nil {
				m.mu.Lock()
				m.failure = err
				m.mu.Unlock()
				m.cancel(err)
				return
			}
		}
	}
}

func (m *leaseMaintainer) stop() error {
	m.stopOnce.Do(func() { close(m.stopCh) })
	<-m.doneCh
	m.cancel(nil)
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.failure
}

File: src/rust/maintained.rs
//! A fail-closed Fiducia + PostgreSQL transaction routine.
//!
//! The legacy [`crate::with_xact_lock`] intentionally preserves its original
//! acquire/work/commit/release behavior. This module provides an opt-in path
//! for long-running work: the Fiducia grant is renewed while PostgreSQL begin,
//! lock acquisition, and user work are in flight, then renewed once more as a
//! mandatory commit-admission check. A failed or malformed renewal rolls the
//! transaction back, so successful return means the caller still held the
//! same fenced authority immediately before commit.

use std::cmp::min;
use std::fmt;
use std::time::Duration;

use sea_orm::{DatabaseTransaction, TransactionTrait};
use tokio::time::{Instant, Interval, MissedTickBehavior, interval_at, sleep_until};

use crate::LockKey;
use crate::coordinated::Guarded;
use crate::error::{LockError, LockErrorKind};
use crate::lease::{AcquireOptions, Lease, LeaseGrant, WorkFuture, release_lost, tag_step};
use crate::pg;
use crate::plan::LockStep;

/// Renewal cadence for [`with_maintained_xact_lock`].
///
/// The interval must be positive and no greater than half of the acquisition
/// TTL. That leaves at least one full interval of safety margin when one tick
/// is delayed by runtime scheduling or a transient slow response.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LeaseMaintenanceOptions {
    pub renew_interval: Duration,
}

impl Default for LeaseMaintenanceOptions {
    fn default() -> Self {
        Self {
            renew_interval: Duration::from_secs(20),
        }
    }
}

impl LeaseMaintenanceOptions {
    pub fn renew_interval(mut self, renew_interval: Duration) -> Self {
        self.renew_interval = renew_interval;
        self
    }

    pub fn renew_interval_ms(&self) -> u64 {
        u64::try_from(self.renew_interval.as_millis()).unwrap_or(u64::MAX)
    }

    /// Validate before either coordination layer is acquired.
    pub fn validate(
        &self,
        key: &LockKey,
        acquire: &AcquireOptions,
        wait: bool,
    ) -> Result<(), LockError> {
        if acquire.ttl.is_zero() {
            return Err(LockError::invalid_plan(
                key,
                "fiducia lease TTL must be greater than zero",
            ));
        }
        if self.renew_interval.is_zero() {
            return Err(LockError::invalid_plan(
                key,
                "fiducia renewal interval must be greater than zero",
            ));
        }
        if self.renew_interval > acquire.ttl / 2 {
            return Err(LockError::invalid_plan(
                key,
                format!(
                    "fiducia renewal interval {} ms is unsafe for TTL {} ms; it must be no greater than half the TTL",
                    self.renew_interval_ms(),
                    acquire.ttl_ms()
                ),
            ));
        }
        if wait && acquire.retry_interval.is_zero() {
            return Err(LockError::invalid_plan(
                key,
                "PostgreSQL advisory-lock retry interval must be greater than zero when wait is enabled",
            ));
        }
        Ok(())
    }
}

/// Run work while holding both a Fiducia lease and
/// `pg_try_advisory_xact_lock`/`pg_advisory_xact_lock` semantics in one
/// PostgreSQL transaction.
///
/// This differs from [`crate::with_xact_lock`] in four safety-critical ways:
///
/// 1. blocking advisory acquisition is implemented as bounded try-lock polling
///    so the Fiducia lease can be renewed and loss can interrupt the wait;
/// 2. the lease is renewed while PostgreSQL begin, lock acquisition, and user
///    work are pending;
/// 3. every renewal must preserve key, holder and fencing token; and
/// 4. a final renewal is required before commit. Any failure rolls back.
///
/// The work future is dropped before rollback when a periodic renewal fails.
/// Therefore all protected database effects must use `guarded.txn`; unrelated
/// external side effects cannot be rolled back by this routine.
pub async fn with_maintained_xact_lock<L, C, T, E, F>(
    key: &LockKey,
    wait: bool,
    acquire: &AcquireOptions,
    maintenance: &LeaseMaintenanceOptions,
    lease: &L,
    db: &C,
    work: F,
) -> Result<T, LockError>
where
    L: Lease + Sync,
    C: TransactionTrait,
    E: fmt::Display,
    F: for<'a> FnOnce(Guarded<'a>) -> WorkFuture<'a, T, E>,
{
    maintenance.validate(key, acquire, wait)?;

    let grant = lease.acquire(key, acquire, wait).await.map_err(|err| {
        tag_step(
            err,
            if wait {
                LockStep::FiduciaAcquire
            } else {
                LockStep::FiduciaTryAcquire
            },
        )
    })?;

    let mut renewals = interval_at(
        Instant::now() + maintenance.renew_interval,
        maintenance.renew_interval,
    );
    renewals.set_missed_tick_behavior(MissedTickBehavior::Delay);

    let begin_result = {
        let begin = db.begin();
        tokio::pin!(begin);
        loop {
            tokio::select! {
                result = &mut begin => break result,
                _ = renewals.tick() => {
                    if let Err(error) = renew_checked(lease, &grant, acquire.ttl).await {
                        return settle_lease(key, lease, &grant, Err(error)).await;
                    }
                }
            }
        }
    };

    let txn = match begin_result {
        Ok(txn) => txn,
        Err(cause) => {
            let inner = LockError::new(LockErrorKind::Database, key, cause.to_string())
                .at(LockStep::PgBegin);
            return settle_lease(key, lease, &grant, Err(inner)).await;
        }
    };

    if let Err(error) =
        acquire_transaction_lock(&txn, key, wait, acquire, lease, &grant, &mut renewals).await
    {
        return rollback_then_settle(key, lease, &grant, txn, error).await;
    }

    let work_result = {
        let work_future = work(Guarded {
            key,
            grant: Some(&grant),
            txn: Some(&txn),
        });
        tokio::pin!(work_future);

        loop {
            tokio::select! {
                result = &mut work_future => {
                    break result.map_err(|cause| LockError::work(key, cause));
                }
                _ = renewals.tick() => {
                    if let Err(error) = renew_checked(lease, &grant, acquire.ttl).await {
                        break Err(error);
                    }
                }
            }
        }
    };

    let value = match work_result {
        Ok(value) => value,
        Err(error) => {
            return rollback_then_settle(key, lease, &grant, txn, error).await;
        }
    };

    // Commit admission: even if the periodic loop renewed one microsecond ago,
    // require one synchronous authority check after work and before commit.
    if let Err(error) = renew_checked(lease, &grant, acquire.ttl).await {
        return rollback_then_settle(key, lease, &grant, txn, error).await;
    }

    let committed = match txn.commit().await {
        Ok(()) => Ok(value),
        Err(cause) => Err(
            LockError::new(LockErrorKind::Database, key, cause.to_string()).at(LockStep::PgCommit),
        ),
    };
    settle_lease(key, lease, &grant, committed).await
}

/// Convenience spelling for the maintained both-layer path with default
/// acquisition and maintenance options.
pub async fn with_maintained_both<L, C, T, E, F>(
    key: &LockKey,
    lease: &L,
    db: &C,
    work: F,
) -> Result<T, LockError>
where
    L: Lease + Sync,
    C: TransactionTrait,
    E: fmt::Display,
    F: for<'a> FnOnce(Guarded<'a>) -> WorkFuture<'a, T, E>,
{
    with_maintained_xact_lock(
        key,
        true,
        &AcquireOptions::default(),
        &LeaseMaintenanceOptions::default(),
        lease,
        db,
        work,
    )
    .await
}

async fn acquire_transaction_lock<L: Lease + Sync>(
    txn: &DatabaseTransaction,
    key: &LockKey,
    wait: bool,
    acquire: &AcquireOptions,
    lease: &L,
    grant: &LeaseGrant,
    renewals: &mut Interval,
) -> Result<(), LockError> {
    let started = Instant::now();

    loop {
        match pg::try_xact_lock(txn, key).await {
            Ok(()) => return Ok(()),
            Err(error) if wait && error.kind == LockErrorKind::Contention => {}
            Err(error) => return Err(error),
        }

        let elapsed = started.elapsed();
        if elapsed >= acquire.wait_timeout {
            return Err(LockError::timeout(
                key,
                LockStep::PgAdvisoryXactLock,
                acquire.wait_timeout_ms(),
            ));
        }

        let remaining = acquire.wait_timeout.saturating_sub(elapsed);
        let retry_at = Instant::now() + min(acquire.retry_interval, remaining);
        loop {
            tokio::select! {
                _ = sleep_until(retry_at) => break,
                _ = renewals.tick() => {
                    renew_checked(lease, grant, acquire.ttl).await?;
                }
            }
        }
    }
}

async fn renew_checked<L: Lease + Sync>(
    lease: &L,
    original: &LeaseGrant,
    ttl: Duration,
) -> Result<LeaseGrant, LockError> {
    let renewed = lease
        .renew(original, ttl)
        .await
        .map_err(|err| tag_step(err, LockStep::FiduciaRenew))?;
    validate_renewed_grant(original, &renewed)?;
    Ok(renewed)
}

fn validate_renewed_grant(original: &LeaseGrant, renewed: &LeaseGrant) -> Result<(), LockError> {
    let mismatch = if renewed.key != original.key {
        Some("key")
    } else if renewed.holder != original.holder {
        Some("holder")
    } else if renewed.fencing_token != original.fencing_token {
        Some("fencing token")
    } else {
        None
    };

    if let Some(field) = mismatch {
        return Err(LockError::new(
            LockErrorKind::LostLease,
            &original.key,
            format!("fiducia renewal changed the grant {field}; fenced authority cannot be proven"),
        )
        .at(LockStep::FiduciaRenew));
    }
    Ok(())
}

async fn rollback_then_settle<L: Lease + Sync, T>(
    key: &LockKey,
    lease: &L,
    grant: &LeaseGrant,
    txn: DatabaseTransaction,
    inner: LockError,
) -> Result<T, LockError> {
    let inner = match txn.rollback().await {
        Ok(()) => inner,
        Err(cause) => LockError::new(LockErrorKind::Database, key, cause.to_string())
            .at(LockStep::PgRollback)
            .after_inner_failure(&inner),
    };
    settle_lease(key, lease, grant, Err(inner)).await
}

async fn settle_lease<L: Lease + Sync, T>(
    key: &LockKey,
    lease: &L,
    grant: &LeaseGrant,
    inner: Result<T, LockError>,
) -> Result<T, LockError> {
    let cleanup = match lease.release(grant).await {
        Ok(true) => None,
        Ok(false) => Some(release_lost(key, grant)),
        Err(error) => Some(tag_step(error, LockStep::FiduciaRelease)),
    };

    match (inner, cleanup) {
        (Ok(value), None) => Ok(value),
        (Err(error), None) => Err(error),
        (Ok(_), Some(cleanup)) => Err(cleanup),
        (Err(inner), Some(cleanup)) => Err(cleanup.after_inner_failure(&inner)),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn grant() -> LeaseGrant {
        LeaseGrant {
            key: LockKey::new("tests/maintained").unwrap(),
            holder: "holder-a".into(),
            fencing_token: 41,
            lease_expires_ms: Some(1_700_000_000_000),
            ttl_ms: 60_000,
        }
    }

    #[test]
    fn maintenance_interval_is_fail_closed() {
        let key = LockKey::new("tests/options").unwrap();
        let acquire = AcquireOptions::default().ttl(Duration::from_millis(100));
        assert!(
            LeaseMaintenanceOptions::default()
                .renew_interval(Duration::from_millis(50))
                .validate(&key, &acquire, true)
                .is_ok()
        );
        for interval in [Duration::ZERO, Duration::from_millis(51)] {
            let error = LeaseMaintenanceOptions::default()
                .renew_interval(interval)
                .validate(&key, &acquire, true)
                .unwrap_err();
            assert_eq!(error.kind, LockErrorKind::InvalidPlan);
        }
    }

    #[test]
    fn renewal_must_preserve_fenced_identity() {
        let original = grant();
        assert!(validate_renewed_grant(&original, &original).is_ok());

        let mut changed = original.clone();
        changed.fencing_token += 1;
        let error = validate_renewed_grant(&original, &changed).unwrap_err();
        assert_eq!(error.kind, LockErrorKind::LostLease);
        assert_eq!(error.step, Some(LockStep::FiduciaRenew));

        let mut changed = original.clone();
        changed.holder = "holder-b".into();
        assert!(validate_renewed_grant(&original, &changed).is_err());

        let mut changed = original.clone();
        changed.key = LockKey::new("tests/other").unwrap();
        assert!(validate_renewed_grant(&original, &changed).is_err());
    }
}
